File: TOP150Review/Valid Parentheses.py
class Solution:
    
    # use stack to solve this problem 
    def isValid(self, s: str) -> bool:
        
        stack = [] 

        ref = {")":"(" , "}":"{" , "]":"["}


        for char in s : 

            if char in ref.values() :   
                stack.append(char) 
            else : 
                
                if stack and stack[-1] == ref[char] : 
                    stack.pop() 
                else : 
                    return False 

        return True if len(stack) == 0 else False                   

File: RUSH/Valid Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */


// stack solution 
var isValid = function(s) {
    
    let refTable = {
        ")":"(" , "]":"[" , "}":"{"
    };

    const stack = [] ;  

    for (let i = 0 ; i < s.length ; i++){

        if (  ["(","[","{"].includes(s[i]) ){
            stack.push(s[i]);
        }
        else{
            if (stack.length === 0 || refTable[s[i]] !== stack[stack.length -1]){
                return false 
            }
            else{
                stack.pop() ; 
            }
        }
    }

    if (stack.length === 0 ) {return true}
    else {return false}
};